pkg/stanza/fileconsumer/matcher/internal/filter/filterregex.go

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package filter // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/matcher/internal/filter"
+import (
+	"regexp"
+)
+
+type filterRegex struct {
+	IncludeRegex              []string
+	ExcludeRegex              []string
+	CapturePathSubstringRegex string
+}
+
+func (fr filterRegex) apply(items []*item) ([]*item, error) {
+	filteredItems := make([]*item, 0, len(items))
+	var errs error
+	var capturedPathSubstring string
+
+	// Compile the regular expression
+	re := regexp.MustCompile(fr.CapturePathSubstringRegex)
+
+	// Find a match and capture the group
+	for _, item := range items {
+		match := re.FindStringSubmatch(item.value)
+
+		if len(match) > 1 {
+			capturedPathSubstring = match[1]
+		} else {
+			continue
+		}
+
+		if len(fr.IncludeRegex) > 0 && len(fr.ExcludeRegex) == 0 {
+
+			val := callIncludeFunc(capturedPathSubstring, fr.IncludeRegex)
+			if val != "" {
+				filteredItems = append(filteredItems, item)
+			}
+
+		} else if len(fr.ExcludeRegex) > 0 && len(fr.IncludeRegex) == 0 {
+
+			val := callExcludeFunc(capturedPathSubstring, fr.ExcludeRegex)
+			if val != "" {
+				filteredItems = append(filteredItems, item)
+			}
+
+		} else {
+			for _, includeregexVar := range fr.IncludeRegex {
+				if includeregexVar != "" {
+					re := regexp.MustCompile(includeregexVar)
```

Comment:
For every item/file name we get, we are doing regexp.MustCompile. Can you keep a list of compiles regexp and then use it in loop?